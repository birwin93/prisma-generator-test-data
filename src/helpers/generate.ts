import { DMMF } from '@prisma/generator-helper'

const TD_OVERRIDE = '@td:'

export function generate(datamodel: DMMF.Datamodel): string {
  return datamodel.models
    .map((model) => generateModel(model, datamodel.enums))
    .join('\n\n')
}

function generateModel(model: DMMF.Model, enums: DMMF.DatamodelEnum[]): string {
  return `export function test${model.name}(): ${model.name} {
        return {
            ${generateModelFields(model, enums)}
        }
    }`
}

function generateModelFields(
  model: DMMF.Model,
  enums: DMMF.DatamodelEnum[],
): string {
  let fields: string[] = []
  for (const field of model.fields) {
    const value = generateFieldValue(field, enums)
    if (value) {
      fields.push(`${field.name}: ${value},`)
    }
  }
  return fields.join('\n')
}

function generateFieldValue(
  field: DMMF.Field,
  enums: DMMF.DatamodelEnum[],
): string | null {
  const docLines = field.documentation?.split('\n') || []
  const tdOverride = docLines
    .find((line) => line.startsWith(TD_OVERRIDE))
    ?.replace(TD_OVERRIDE, '')

  console.log(docLines, tdOverride)

  // If there is an explict td override, use it
  if (tdOverride) {
    return tdOverrideValue(tdOverride)
  }

  // If field is an object (aka relation)
  // or is optional, return null
  if (field.kind === 'object' || field.isRequired == false) {
    return null
  }

  // If field is a scalar, check type and return appropriate value
  if (field.kind == 'scalar') {
    if (field.isId) {
      if (field.type == 'String') {
        return 'faker.string.uuid()'
      } else {
        return 'faker.number.int()'
      }
    } else if (field.type == 'String') {
      if (field.default) {
        return `'${field.default}'`
      } else {
        return 'faker.lorem.word(10)'
      }
    } else if (field.type == 'Date' || field.type == 'DateTime') {
      return 'new Date()'
    } else if (field.type == 'Int') {
      return 'faker.number.int()'
    }
  } else if (field.kind == 'enum') {
    const enumName = field.type
    const enumValues = enums.find((it) => it.name === enumName)?.values || []
    if (enumValues.length === 0) {
      throw new Error(`Could not find enum values for enum ${enumName}`)
    } else {
      return `${enumName}.${enumValues[0].name}`
    }
  }

  return `${field}`
}

function tdOverrideValue(override: string): string {
  switch (override) {
    case 'email':
      return 'faker.internet.email()'
    case 'phone':
      return 'faker.phone.number()'
    case 'firstName':
      return 'faker.person.firstName()'
    case 'lastName':
      return 'faker.person.lastName()'
    case 'fullName':
      return 'faker.person.findName()'
    case 'numberString':
      return 'faker.string.numeric(8)'
    case 'bigint':
      return 'faker.number.int()'
  }

  return override
}
