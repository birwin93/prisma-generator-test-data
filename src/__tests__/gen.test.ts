import { getSampleDMMF } from './__fixtures__/getSampleDMMF'
import { generate } from '../helpers/generate'

test('generate', async () => {
  const sampleDMMF = await getSampleDMMF()
  expect(generate(sampleDMMF.datamodel)).toMatchSnapshot("test_snapshot")
})
