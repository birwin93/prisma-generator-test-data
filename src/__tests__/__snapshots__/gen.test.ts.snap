// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate: test_snapshot 1`] = `
"export function testUser(): User {
        return {
            id: faker.string.uuid(),
rewardsId: faker.string.numeric(8),
ezLinksContactId: faker.string.numeric(8),
email: faker.internet.email(),
role: Role.USER,
firstName: faker.person.firstName(),
lastName: faker.person.lastName(),
zipCode: '00000',
createdAt: new Date(),
updatedAt: new Date(),
        }
    }

export function testCourse(): Course {
        return {
            id: faker.string.uuid(),
ezlinksId: faker.number.int(),
troonFacilityId: faker.number.int(),
name: courseName,
facilityName: facilityName,
phone: faker.phone.number(),
latitude: latitude,
longitude: longitude,
holeCount: 18,
createdAt: new Date(),
updatedAt: new Date(),
        }
    }

export function testReservation(): Reservation {
        return {
            id: faker.string.uuid(),
teeTimeId: faker.lorem.word(10),
courseId: faker.lorem.word(10),
state: ReservationState.ACTIVE,
playerCount: faker.number.int(),
dayTimeString: faker.lorem.word(10),
ownerId: faker.lorem.word(10),
createdAt: new Date(),
updatedAt: new Date(),
        }
    }

export function testReservationUser(): ReservationUser {
        return {
            id: faker.string.uuid(),
state: ReservationUserState.EMPTY,
reservationId: faker.lorem.word(10),
originalCost: [object Object],
createdAt: new Date(),
updatedAt: new Date(),
        }
    }"
`;
